runtime/stdlib.js

```diff
@@ -48,17 +43,22 @@ function caml_call_gen(f, args) {
       f = f.fun;
       continue;
     }
+    // TODO: This can happen with over-application. Should we fail here ?
+    if (typeof f !== "function") return f;
 
-    var n = f.length;
-    var argsLen = args.length;
-    var d = n - argsLen;
+    var n = f.length | 0;
+    var argsLen = args.length | 0;
+    var d = n - argsLen | 0;
 
     if (d == 0) {
       return f.apply(null, args);
     }
     else if (d < 0) {
       if (!args_copied) {
-        args = args.slice();
+        if(!args.slice)
+          args = Array.prototype.slice.call(args);
+        else
+          args = args.slice();
```

Comment:
Is there an advantage to the conditional? Otherwise I'd expect this (which works for both arrays and `Arguments`):

```suggestion
        args = Array.prototype.slice.call(args);
```

Reply:
I don't know if there are any differences in perf. @TyOverby, any idea ?